core/src/main/scala/kafka/controller/ZookeeperClient.scala

```diff
@@ -36,13 +36,18 @@ import org.apache.zookeeper.{CreateMode, WatchedEvent, Watcher, ZooKeeper}
   * @param connectionTimeoutMs connection timeout in milliseconds
   * @param stateChangeHandler state change handler callbacks called by the underlying zookeeper client's EventThread.
   */
-class ZookeeperClient(connectString: String, sessionTimeoutMs: Int, connectionTimeoutMs: Int, stateChangeHandler: StateChangeHandler) extends Logging {
+class ZookeeperClient(connectString: String,
+                      sessionTimeoutMs: Int,
+                      connectionTimeoutMs: Int,
+                      stateChangeHandler: StateChangeHandler,
+                      maxInFlightRequests: Int = Int.MaxValue) extends Logging {
```

Comment:
Where is the lower limit specified ?

Reply:
Hi @tedyu. What do you mean by lower limit?

Reply:
By default, there is no limit.
Who would be enforcing the limit ?

Reply:
I'm still not sure I understand your question.

The client's maxInFlightRequests is configurable.

If you want the feature, you'd set the limit when instantiating the client.
If you don't want the feature, it's set to `Int.MaxValue` by default which effectively means there is no limit.

The actual limit is enforced through the `inFlightRequests` semaphore which increments when a request goes over the wire and decrements when we get a response back.

Reply:
Is it possible to add unit test which imposes lower limit and verifies that this works ?

Thanks for your patience.

Reply:
The release of semaphore and decrement of countDownLatch are paired.
Consider creating a method which does these two operations.

Reply:
Could we get rid of the default?